Firebase/Database/Core/FServerValues.m

```diff
@@ -20,37 +20,118 @@
+        return @(r);
+    }
+
+  return [NSNumber numberWithDouble:(double)x + y];
+}
+
+@interface FServerValues ()
++ (id)resolveScalarServerOp:(NSString *)op
+           withServerValues:(NSDictionary *)serverValues;
++ (id)resolveComplexServerOp:(NSDictionary *)op
+                withExisting:(id<FNode>)existing
+                serverValues:(NSDictionary *)serverValues;
+@end
+
 @implementation FServerValues
 
 + (NSDictionary *)generateServerValues:(id<FClock>)clock {
     long long millis = (long long)([clock currentTime] * 1000);
-    return @{@"timestamp" : [NSNumber numberWithLongLong:millis]};
+    return @{kTimestamp : [NSNumber numberWithLongLong:millis]};
 }
 
 + (id)resolveDeferredValue:(id)val
-          withServerValues:(NSDictionary *)serverValues {
-    if ([val isKindOfClass:[NSDictionary class]]) {
-        NSDictionary *dict = val;
-        if (dict[kServerValueSubKey] != nil) {
-            NSString *serverValueType = [dict objectForKey:kServerValueSubKey];
-            if (serverValues[serverValueType] != nil) {
-                return [serverValues objectForKey:serverValueType];
-            } else {
-                // TODO: Throw unrecognizedServerValue error here
-            }
-        }
+              withExisting:(id<FNode>)existing
+              serverValues:(NSDictionary *)serverValues {
+    if (![val isKindOfClass:[NSDictionary class]]) {
+        return val;
+    }
+    NSDictionary *dict = val;
+    id op = dict[kServerValueSubKey];
+
+    if (op == nil) {
+        return val;
+    } else if ([op isKindOfClass:NSString.class]) {
+        return [FServerValues resolveScalarServerOp:op
+                                   withServerValues:serverValues];
+    } else if ([op isKindOfClass:NSDictionary.class]) {
+        return [FServerValues resolveComplexServerOp:op
+                                        withExisting:existing
+                                        serverValues:serverValues];
     }
     return val;
 }
 
++ (id)resolveScalarServerOp:(NSString *)op
+           withServerValues:(NSDictionary *)serverValues {
+    return serverValues[op];
+}
+
++ (id)resolveComplexServerOp:(NSDictionary *)op
+                withExisting:(id<FNode>)existing
+                serverValues:(NSDictionary *)serverValues {
+    // Only increment is supported as of now
+    if (op[kIncrement] == nil) {
+        return nil;
+    }
+
+    // Incrementing a non-number sets the value to the incremented amount
+    NSNumber *delta = op[kIncrement];
+    if (![existing isLeafNode]) {
+        return delta;
+    }
+    FLeafNode *existingLeaf = existing;
+    if (![existingLeaf.value isKindOfClass:NSNumber.class]) {
+        return delta;
+    }
+
+    NSNumber *existingNum = existingLeaf.value;
+    BOOL incrLong = longCanHold(delta);
+    BOOL baseLong = longCanHold(existingNum);
+
+    if (incrLong && baseLong) {
+      return overflowAwareAddLong(delta.longValue, existingNum.longValue);
```

Comment:
I'm also not a huge fan of this method name and would suggest to inline.

Optional.

Reply:
inlining also simplifies, so yay